File: tutorials/dynamic-receiving-with-mpi-probe-and-mpi-status/code/check_status.c
// Author: Wes Kendall
// Copyright 2011 www.mpitutorial.com
// This code is provided freely with the tutorials on mpitutorial.com. Feel
// free to modify it for your own use. Any distribution of the code must
// either provide a link to www.mpitutorial.com or keep this header intact.
//
// Example of checking the MPI_Status object from an MPI_Recv call
//
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char** argv) {
  MPI_Init(NULL, NULL);

  int world_size;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
  if (world_size != 2) {
    fprintf(stderr, "Must use two processes for this example\n");
    MPI_Abort(MPI_COMM_WORLD, 1);
  }
  int world_rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);

  const int MAX_NUMBERS = 100;
  int numbers[MAX_NUMBERS];
  int number_amount;
  if (world_rank == 0) {
    // Pick a random amount of integers to send to process one
    srand(time(NULL));
    number_amount = (rand() / (float)RAND_MAX) * MAX_NUMBERS;
    // Send the amount of integers to process one
    MPI_Send(numbers, number_amount, MPI_INT, 1, 0, MPI_COMM_WORLD);
    printf("0 sent %d numbers to 1\n", number_amount);
  } else if (world_rank == 1) {
    MPI_Status status;
    // Receive at most MAX_NUMBERS from process zero
    MPI_Recv(numbers, MAX_NUMBERS, MPI_INT, 0, 0, MPI_COMM_WORLD, &status);
    // After receiving the message, check the status to determine how many
    // numbers were actually received
    MPI_Get_count(&status, MPI_INT, &number_amount);
    // Print off the amount of numbers, and also print additional information
    // in the status object
    printf("1 received %d numbers from 0. Message source = %d, tag = %d\n",
           number_amount, status.MPI_SOURCE, status.MPI_TAG);
  }
  MPI_Barrier(MPI_COMM_WORLD);
  MPI_Finalize();
}

File: tutorials/mpi-fortran/code/check_status.f90
program check_status
  use mpi_f08

  implicit none

  integer            :: world_rank, world_size
  integer, parameter :: MAX_NUMBERS=100
  integer            :: numbers(MAX_NUMBERS)
  integer            :: number_amount
  type(MPI_Status)   :: recv_status

  real :: r

  call MPI_INIT()
  call MPI_COMM_SIZE(MPI_COMM_WORLD, world_size)
  call MPI_COMM_RANK(MPI_COMM_WORLD, world_rank)


  if (world_rank .eq. 0) then
    ! Pick a random amount of integers to send to process one
    call random_seed()

    call random_number(r)
    number_amount = int(r * real(MAX_NUMBERS))
    ! Send the amount of integers to process one
    call MPI_SEND(numbers, number_amount, MPI_INT, 1, 9, &
                  MPI_COMM_WORLD)
    print '("0 sent ", I0, " numbers to 1")', number_amount
  else if (world_rank .eq. 1) then
    ! Receive at most MAX_NUMBERS from process zero
    call MPI_RECV(numbers, MAX_NUMBERS, MPI_INT, 0, 9, &
                  MPI_COMM_WORLD, recv_status)
    ! After receiving the message, check the status to determine how many
    ! numbers were actually received
    call MPI_Get_count(recv_status, MPI_INT, number_amount)
    ! Print off the amount of numbers, and also print additional information
    ! in the status object
    print '("1 received ", I0, " numbers from 0. Message source = ", I0, ", tag = ", I0)', &
      number_amount , recv_status%MPI_SOURCE , recv_status%MPI_TAG
  end if

  call MPI_Barrier(MPI_COMM_WORLD)

  call MPI_FINALIZE()

end program
